src/Stripe.Tests.XUnit/coupons_new/coupons.cs

```diff
@@ -0,0 +1,116 @@
+using System.Collections.Generic;
+using System.Linq;
+using FluentAssertions;
+using Xunit;
+
+namespace Stripe.Tests.Xunit
+{
+    public class coupon_test
+    {
+        private string previousApiBase;
+        private readonly string resourceId = "co_123";
+        private StripeCouponService service;
+
+        public coupon_test()
+        {
+            this.service = new StripeCouponService(Cache.ApiKey);
+            this.previousApiBase = StripeConfiguration.GetApiBase();
+            StripeConfiguration.SetApiBase("http://localhost:12111/v1");
+        }
+
+        ~coupon_test()
+        {
+            StripeConfiguration.SetApiBase(this.previousApiBase);
```

Comment:
I don't think this works. With xUnit, if you have cleanup code, your test class must implement `IDisposable` and the cleanup code must be placed in a `Dispose()` method. Cf. https://xunit.github.io/docs/shared-context.

Reply:
hmmm weirdly it definitely worked locally but I'll investigate further in that case

Reply:
I think that either will often work in practice, but the difference is that unlike destructors in C/C++, finalizers in C# are not invoked deterministically — they'll be invoked by the GC only when it deems it's the right time to clean up the instance. For this sort of test teardown, that might produce some unexpected results when you're running a number of tests together that are all modifying this sort of global configuration. ([See docs on finalizers for details](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/destructors).)

As OB suggests, you probably want `Dispose` instead because XUnit can invoke that explicitly as soon as this suite has finished running.

Reply:
The destructor calls .Finalize() in .net... that's why this got cleaned up when the garbage collector was finished with it. With `IDisposable`, you have more control. You can call Dispose() or implement your code in a `using` block. The Finalizer (.Finalize() or ~) is going to run when the object gets destroyed, but unfortunately you don't know exactly when that will occur.